Crash in on restore when ViewStack parameter contains a Serializable object

I am getting the following crash when restoring if the ViewStack contains parameters with a Serializable object.

```
Process: com.solera.defragsample, PID: 10308
java.lang.RuntimeException: Parcelable encountered ClassNotFoundException reading a Serializable object (name = com.solera.defragsample.BreakdownPresenter$Test)
    at android.os.Parcel.readSerializable(Parcel.java:2630)
    at android.os.Parcel.readValue(Parcel.java:2416)
    at android.os.Parcel.readArrayMapInternal(Parcel.java:2732)
    at android.os.BaseBundle.unparcel(BaseBundle.java:269)
    at android.os.BaseBundle.getInt(BaseBundle.java:854)
    at com.solera.defragsample.BreakdownPresenter.onTakeView(BreakdownPresenter.java:44)
    at com.solera.defragsample.Presenter.takeView(Presenter.java:40)
    at com.solera.defragsample.BreakdownView.onAttachedToWindow(BreakdownView.java:53)
    at android.view.View.dispatchAttachedToWindow(View.java:15509)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2916)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923)
    at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1526)
    at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1254)
    at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6337)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874)
    at android.view.Choreographer.doCallbacks(Choreographer.java:686)
    at android.view.Choreographer.doFrame(Choreographer.java:621)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
    at android.os.Handler.handleCallback(Handler.java:751)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:154)
    at android.app.ActivityThread.main(ActivityThread.java:6119)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
 Caused by: java.lang.ClassNotFoundException: com.solera.defragsample.BreakdownPresenter$Test
    at java.lang.Class.classForName(Native Method)
    at java.lang.Class.forName(Class.java:400)
    at android.os.Parcel$2.resolveClass(Parcel.java:2616)
    at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
    at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1772)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
    at java.io.ObjectInputStream.readObject(ObjectInputStream.java:373)
    at android.os.Parcel.readSerializable(Parcel.java:2624)
    at android.os.Parcel.readValue(Parcel.java:2416) 
    at android.os.Parcel.readArrayMapInternal(Parcel.java:2732) 
    at android.os.BaseBundle.unparcel(BaseBundle.java:269) 
    at android.os.BaseBundle.getInt(BaseBundle.java:854) 
    at com.solera.defragsample.BreakdownPresenter.onTakeView(BreakdownPresenter.java:44) 
    at com.solera.defragsample.Presenter.takeView(Presenter.java:40) 
    at com.solera.defragsample.BreakdownView.onAttachedToWindow(BreakdownView.java:53) 
    at android.view.View.dispatchAttachedToWindow(View.java:15509) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2916) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1526) 
    at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1254) 
    at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6337) 
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874) 
    at android.view.Choreographer.doCallbacks(Choreographer.java:686) 
    at android.view.Choreographer.doFrame(Choreographer.java:621) 
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860) 
    at android.os.Handler.handleCallback(Handler.java:751) 
    at android.os.Handler.dispatchMessage(Handler.java:95) 
    at android.os.Looper.loop(Looper.java:154) 
    at android.app.ActivityThread.main(ActivityThread.java:6119) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776) 
 Caused by: java.lang.ClassNotFoundException: com.solera.defragsample.BreakdownPresenter$Test
    at java.lang.Class.classForName(Native Method)
    at java.lang.BootClassLoader.findClass(ClassLoader.java:1346)
    at java.lang.BootClassLoader.loadClass(ClassLoader.java:1406)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
    at java.lang.Class.classForName(Native Method) 
    at java.lang.Class.forName(Class.java:400) 
    at android.os.Parcel$2.resolveClass(Parcel.java:2616) 
    at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613) 
    at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518) 
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1772) 
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351) 
    at java.io.ObjectInputStream.readObject(ObjectInputStream.java:373) 
    at android.os.Parcel.readSerializable(Parcel.java:2624) 
    at android.os.Parcel.readValue(Parcel.java:2416) 
    at android.os.Parcel.readArrayMapInternal(Parcel.java:2732) 
    at android.os.BaseBundle.unparcel(BaseBundle.java:269) 
    at android.os.BaseBundle.getInt(BaseBundle.java:854) 
    at com.solera.defragsample.BreakdownPresenter.onTakeView(BreakdownPresenter.java:44) 
    at com.solera.defragsample.Presenter.takeView(Presenter.java:40) 
    at com.solera.defragsample.BreakdownView.onAttachedToWindow(BreakdownView.java:53) 
    at android.view.View.dispatchAttachedToWindow(View.java:15509) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2916) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:2923) 
    at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1526) 
    at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1254) 
    at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6337) 
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874) 
    at android.view.Choreographer.doCallbacks(Choreographer.java:686) 
    at android.view.Choreographer.doFrame(Choreographer.java:621) 
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860) 
    at android.os.Handler.handleCallback(Handler.java:751) 
    at android.os.Handler.dispatchMessage(Handler.java:95) 
    at android.os.Looper.loop(Looper.java:154) 
    at android.app.ActivityThread.main(ActivityThread.java:6119) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776) 
 Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
```

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -2,13 +2,13 @@ apply plugin: 'com.android.application'
 apply plugin: 'android-apt'
 
 android {
-  compileSdkVersion 23
-  buildToolsVersion '23.0.3'
+  compileSdkVersion 25
+  buildToolsVersion '25.0.1'
 
   defaultConfig {
     applicationId "com.solera.defragsample"
     minSdkVersion 16
-    targetSdkVersion 23
+    targetSdkVersion 25
     versionCode 1
     versionName "1.0"
   }

--- a/build.gradle
+++ b/build.gradle
@@ -5,7 +5,7 @@ buildscript {
     jcenter()
   }
   dependencies {
-    classpath 'com.android.tools.build:gradle:2.2.2'
+    classpath 'com.android.tools.build:gradle:2.2.3'
     classpath 'com.neenbedankt.gradle.plugins:android-apt:1.8'
     classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.6'
     classpath 'com.github.dcendents:android-maven-gradle-plugin:1.4.1'

--- a/defrag/build.gradle
+++ b/defrag/build.gradle
@@ -8,12 +8,12 @@ version = '1.1.1'
 
 
 android {
-    compileSdkVersion 23
-    buildToolsVersion "23.0.3"
+    compileSdkVersion 25
+    buildToolsVersion "25.0.1"
 
     defaultConfig {
         minSdkVersion 16
-        targetSdkVersion 23
+        targetSdkVersion 25
         versionCode 2
         versionName "${version}"
     }
@@ -31,12 +31,6 @@ dependencies {
 
     compile "com.android.support:appcompat-v7:${android_support_lib_version}"
     compile "com.android.support:support-annotations:${android_support_lib_version}"
-
-    // Auto value
-    provided "com.google.auto.value:auto-value:${auto_value_lib_version}"
-    apt "com.google.auto.value:auto-value:${auto_value_lib_version}"
-    apt "com.ryanharter.auto.value:auto-value-parcel:0.2.5"
-    compile "com.ryanharter.auto.value:auto-value-parcel-adapter:0.2.5"
 }
 
 def siteUrl = 'https://github.com/R3PI/Defrag'

--- a/defrag/src/main/java/com/solera/defrag/TraversalAnimation.java
+++ b/defrag/src/main/java/com/solera/defrag/TraversalAnimation.java
@@ -3,23 +3,56 @@ package com.solera.defrag;
 import android.animation.Animator;
 import android.support.annotation.IntDef;
 import android.support.annotation.NonNull;
-import com.google.auto.value.AutoValue;
 import java.lang.annotation.Retention;
 
 import static java.lang.annotation.RetentionPolicy.SOURCE;
 
-@AutoValue public abstract class TraversalAnimation {
+public class TraversalAnimation {
 	public static final int ABOVE = 0;
 	public static final int BELOW = 1;
+	private final Animator mAnimator;
+	private final int mDrawOrder;
 
 	@NonNull public static TraversalAnimation newInstance(@NonNull Animator animator,
 			@AnimateInDrawOrder int drawOrder) {
-		return new AutoValue_TraversalAnimation(animator, drawOrder);
+		return new TraversalAnimation(animator, drawOrder);
 	}
 
-	@NonNull abstract Animator animator();
+	private TraversalAnimation(@NonNull Animator animator, @AnimateInDrawOrder int drawOrder) {
+		mAnimator = animator;
+		mDrawOrder = drawOrder;
+	}
 
-	@AnimateInDrawOrder abstract int drawOrder();
+	@NonNull Animator animator() {
+		return mAnimator;
+	}
+
+	@AnimateInDrawOrder int drawOrder() {
+		return mDrawOrder;
+	}
+
+	@Override public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) return false;
+
+		TraversalAnimation that = (TraversalAnimation) o;
+
+		if (mDrawOrder != that.mDrawOrder) return false;
+		return mAnimator != null ? mAnimator.equals(that.mAnimator) : that.mAnimator == null;
+	}
+
+	@Override public int hashCode() {
+		int result = mAnimator != null ? mAnimator.hashCode() : 0;
+		result = 31 * result + mDrawOrder;
+		return result;
+	}
+
+	@Override public String toString() {
+		return "TraversalAnimation{" +
+				"mAnimator=" + mAnimator +
+				", mDrawOrder=" + mDrawOrder +
+				'}';
+	}
 
 	@Retention(SOURCE) @IntDef({ ABOVE, BELOW }) public @interface AnimateInDrawOrder {
 	}

--- a/defrag/src/main/java/com/solera/defrag/ViewStack.java
+++ b/defrag/src/main/java/com/solera/defrag/ViewStack.java
@@ -33,9 +33,6 @@ import android.util.SparseArray;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.widget.FrameLayout;
-import com.google.auto.value.AutoValue;
-import com.ryanharter.auto.value.parcel.ParcelAdapter;
-import com.ryanharter.auto.value.parcel.TypeAdapter;
 import java.io.Serializable;
 import java.lang.ref.WeakReference;
 import java.util.ArrayDeque;
@@ -370,7 +367,9 @@ public class ViewStack extends FrameLayout {
 		while (viewStackEntryIterator.hasNext()) {
 			final ViewStackEntry viewStackEntry = viewStackEntryIterator.next();
 			if (view == viewStackEntry.mViewReference.get()) {
-				return viewStackEntry.mParameters;
+				Bundle bundle = viewStackEntry.mParameters;
+				bundle.setClassLoader(view.getClass().getClassLoader());
+				return bundle;
 			}
 		}
 		return null;
@@ -436,8 +435,7 @@ public class ViewStack extends FrameLayout {
 		final SaveState parcelable = (SaveState) state;
 		for (SaveStateEntry entry : parcelable.stack()) {
 			//we have to cast to SparseArray as we can't serialize a SparseArray<Parcelable>
-			viewStack.add(
-					new ViewStackEntry(entry.layout(), entry.parameters(), (SparseArray) entry.viewState()));
+			viewStack.add(new ViewStackEntry(entry.layout(), entry.parameters(), entry.viewState()));
 		}
 		if (!viewStack.isEmpty()) {
 			addView(viewStack.peek().getView());
@@ -491,41 +489,182 @@ public class ViewStack extends FrameLayout {
 		return parameterBundle;
 	}
 
-	@AutoValue static abstract class SaveState implements Parcelable {
+	static class SaveState implements Parcelable {
+		private List<SaveStateEntry> mStack;
+		private Parcelable mSuperState;
+
 		static SaveState newInstance(@NonNull ViewStack viewstack, @NonNull Parcelable superState) {
 			List<SaveStateEntry> stack = new ArrayList<>(viewstack.getViewCount());
 			for (ViewStackEntry entry : viewstack.viewStack) {
 				stack.add(SaveStateEntry.newInstance(entry.mLayout, entry.mParameters, entry.mViewState));
 			}
-			return new AutoValue_ViewStack_SaveState(stack, superState);
+			return new SaveState(stack, superState);
 		}
 
-		@NonNull abstract List<SaveStateEntry> stack();
+		SaveState(@NonNull List<SaveStateEntry> stack, @NonNull Parcelable superState) {
+			mStack = stack;
+			mSuperState = superState;
+		}
 
-		@NonNull abstract Parcelable superState();
+		SaveState(Parcel in) {
+			if (in.readByte() == 0x01) {
+				mStack = new ArrayList<SaveStateEntry>();
+				in.readList(mStack, SaveStateEntry.class.getClassLoader());
+			} else {
+				mStack = null;
+			}
+			mSuperState = in.readParcelable(Parcelable.class.getClassLoader());
+		}
+
+		@NonNull List<SaveStateEntry> stack() {
+			return mStack;
+		}
+
+		@NonNull Parcelable superState() {
+			return mSuperState;
+		}
+
+		@Override public int describeContents() {
+			return 0;
+		}
+
+		@Override public void writeToParcel(Parcel dest, int flags) {
+			if (mStack == null) {
+				dest.writeByte((byte) (0x00));
+			} else {
+				dest.writeByte((byte) (0x01));
+				dest.writeList(mStack);
+			}
+			dest.writeParcelable(mSuperState, PARCELABLE_WRITE_RETURN_VALUE);
+		}
+
+		@Override public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+
+			SaveState saveState = (SaveState) o;
+
+			if (mStack != null ? !mStack.equals(saveState.mStack) : saveState.mStack != null)
+				return false;
+			return mSuperState != null ? mSuperState.equals(saveState.mSuperState)
+					: saveState.mSuperState == null;
+		}
+
+		@Override public int hashCode() {
+			int result = mStack != null ? mStack.hashCode() : 0;
+			result = 31 * result + (mSuperState != null ? mSuperState.hashCode() : 0);
+			return result;
+		}
+
+		@Override public String toString() {
+			return "SaveState{" +
+					"mStack=" + mStack +
+					", mSuperState=" + mSuperState +
+					'}';
+		}
+
+
+		@SuppressWarnings("unused")
+		public static final Parcelable.Creator<SaveState> CREATOR = new Parcelable.Creator<SaveState>() {
+			@Override
+			public SaveState createFromParcel(Parcel in) {
+				return new SaveState(in);
+			}
+
+			@Override
+			public SaveState[] newArray(int size) {
+				return new SaveState[size];
+			}
+		};
 	}
 
-	@AutoValue static abstract class SaveStateEntry implements Parcelable {
-		static SaveStateEntry newInstance(int layout, @Nullable Bundle parameters,
+	static class SaveStateEntry implements Parcelable {
+		@LayoutRes private final int mLayout;
+		private final Bundle mParameters;
+		private final SparseArray<Parcelable> mViewState;
+
+		static SaveStateEntry newInstance(@LayoutRes int layout, @Nullable Bundle parameters,
 				@Nullable SparseArray<Parcelable> viewState) {
-			return new AutoValue_ViewStack_SaveStateEntry(layout, parameters, viewState);
+			return new SaveStateEntry(layout, parameters, viewState);
 		}
 
-		@LayoutRes abstract int layout();
-
-		@Nullable abstract Bundle parameters();
-
-		@ParcelAdapter(SaveStateEntryViewStateAdapter.class) @Nullable abstract SparseArray<Parcelable> viewState();
-	}
-
-	static class SaveStateEntryViewStateAdapter implements TypeAdapter<SparseArray<Parcelable>> {
-		public SparseArray<Parcelable> fromParcel(Parcel in) {
-			return in.readSparseArray(AutoValue_ViewStack_SaveStateEntry.class.getClassLoader());
+		SaveStateEntry(@LayoutRes int layout, @Nullable Bundle parameters,
+				@Nullable SparseArray<Parcelable> viewState) {
+			mLayout = layout;
+			mParameters = parameters;
+			mViewState = viewState;
 		}
 
-		public void toParcel(SparseArray<Parcelable> value, Parcel dest) {
-			dest.writeSparseArray((SparseArray) value);
+		SaveStateEntry(Parcel in) {
+			mLayout = in.readInt();
+			mParameters = in.readBundle();
+			mViewState = (SparseArray) in.readValue(SparseArray.class.getClassLoader());
 		}
+
+		@LayoutRes int layout() {
+			return mLayout;
+		}
+
+		@Nullable Bundle parameters() {
+			return mParameters;
+		}
+
+		@Nullable SparseArray<Parcelable> viewState() {
+			return mViewState;
+		}
+
+		@Override public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+
+			SaveStateEntry that = (SaveStateEntry) o;
+
+			if (mLayout != that.mLayout) return false;
+			if (mParameters != null ? !mParameters.equals(that.mParameters) : that.mParameters != null) {
+				return false;
+			}
+			return mViewState != null ? mViewState.equals(that.mViewState) : that.mViewState == null;
+		}
+
+		@Override public int hashCode() {
+			int result = mLayout;
+			result = 31 * result + (mParameters != null ? mParameters.hashCode() : 0);
+			result = 31 * result + (mViewState != null ? mViewState.hashCode() : 0);
+			return result;
+		}
+
+		@Override public String toString() {
+			return "SaveStateEntry{" +
+					"mLayout=" + mLayout +
+					", mParameters=" + mParameters +
+					", mViewState=" + mViewState +
+					'}';
+		}
+
+		@Override
+		public int describeContents() {
+			return 0;
+		}
+
+		@Override
+		public void writeToParcel(Parcel dest, int flags) {
+			dest.writeInt(mLayout);
+			dest.writeBundle(mParameters);
+			dest.writeValue(mViewState);
+		}
+
+		@SuppressWarnings("unused")
+		public static final Parcelable.Creator<SaveStateEntry> CREATOR = new Parcelable.Creator<SaveStateEntry>() {
+			@Override
+			public SaveStateEntry createFromParcel(Parcel in) {
+				return new SaveStateEntry(in);
+			}
+
+			@Override
+			public SaveStateEntry[] newArray(int size) {
+				return new SaveStateEntry[size];
+			}
+		};
 	}
 
 	private class ViewStackEntry {
@@ -546,7 +685,7 @@ public class ViewStack extends FrameLayout {
 		}
 
 		void saveState(@NonNull View view) {
-			final SparseArray<Parcelable> parcelableSparseArray = new SparseArray<>();
+			final SparseArray<Parcelable> parcelableSparseArray = new SparseArray<Parcelable>();
 			view.saveHierarchyState(parcelableSparseArray);
 			mViewState = parcelableSparseArray;
 		}

--- a/gradle.properties
+++ b/gradle.properties
@@ -18,7 +18,7 @@
 # org.gradle.parallel=true
 
 # Google libs
-android_support_lib_version = 23.4.0
+android_support_lib_version = 25.1.0
 
 #Auto value
 auto_value_lib_version = 1.3

